gwpy/table/gravityspy.py

```diff
@@ -69,187 +138,129 @@ def download(self, **kwargs):
-                                 'trainingset* table?')
-
-        # If someone wants labelled samples they are
-        # Definitely asking for the training set but
-        # may hve forgotten
-        if labelled_samples and not training_set:
-            training_set = 1
-
-        # Let us check what columns are needed
-        cols_for_download = ['url1', 'url2', 'url3', 'url4']
-        cols_for_download = [cols_for_download[idx] for idx in duration_idx]
-        cols_for_download_ext = ['ml_label', 'sample_type',
-                                 'ifo', 'gravityspy_id']
-
-        if not training_set:
-            images_db['ml_label'] = ''
-        if not labelled_samples:
-            images_db['sample_type'] = ''
-
-        if not os.path.isdir(download_location):
-            os.makedirs(download_location)
-
-        if training_set:
-            for ilabel in images_db.ml_label.unique():
-                if labelled_samples:
-                    for itype in images_db.sample_type.unique():
-                        if not os.path.isdir(os.path.join(
-                                             download_location,
-                                             ilabel, itype)):
-                            os.makedirs(os.path.join(download_location,
-                                        ilabel, itype))
-                else:
-                    if not os.path.isdir(os.path.join(download_location,
-                                                      ilabel)):
-                        os.makedirs(os.path.join(download_location,
-                                                 ilabel))
-
-        images_for_download = images_db[cols_for_download]
-        images = images_for_download.values.flatten()
-        images_for_download_ext = images_db[cols_for_download_ext]
-        duration = np.atleast_2d(
-            duration_values.repeat(
-                len(images_for_download_ext),
-                0,
-            ).flatten(),
-        ).T
-        images_for_download_ext = images_for_download_ext.values.repeat(
-            len(cols_for_download),
-            0,
-        )
-        images_for_for_download_path = np.array([[download_location]]).repeat(
-            len(images_for_download_ext),
-            0,
-        )
-        images = np.hstack((np.atleast_2d(images).T,
-                           images_for_download_ext, duration,
-                           images_for_for_download_path))
-
-        # calculate maximum number of processes
-        nproc = min(kwargs.pop('nproc', 1), len(images))
-
-        # read files
-        output = mp_utils.multiprocess_with_queues(
+        download_path = Path(download_path)
+
+        # work out which columns contain URLs
+        url_columns = list(filter(URL_COLUMN.match, self.colnames))
+
+        # construct a download target location for each requested image
+        urls = dict()
+        for row in self:
+            if not self["url1"]:  # skip rows without URLs
```

Comment:
Hey Duncan,

thanks for doing this! just need to change this line 155 to

```
            if row["url1"] == "?":  # skip rows without URLs
                continue
```

Reply:
This will let the download skip any rows that do not have a public url associated with them in the databases/table.

Reply:
@scottcoughlin2014, do you mean

```suggestion
            if not row["url1"]:  # skip rows without URLs
```

to just fix the typo, or 

```suggestion
            if row["url1"] == "?":  # skip rows without URLs
```

to actually change the refence 'no url' value? If the latter should the test actually be

```suggestion
            if self["url1"] in {None, "", "?"}:  # skip rows without URLs
```

to catch actual empty values?

Reply:
Hey @duncanmmacleod , I think it would be most robust to be

`            if row["url1"] in {None, "", "?"}:  # skip rows without URLs`

Reply:
This change has been made.